src/unistd.rs

```diff
@@ -2136,3 +2136,201 @@ mod setres {
         Errno::result(res).map(drop)
     }
 }
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Passwd {
+    /// username
+    pub name: CString,
+    /// user password (probably encrypted)
+    pub passwd: CString,
+    /// user ID
+    pub uid: Uid,
+    /// group ID
+    pub gid: Gid,
+    #[cfg(not(target_os = "android"))]
+    /// user information
+    pub gecos: CString,
+    /// home directory
+    pub dir: CString,
+    /// path to shell
+    pub shell: CString,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// login class
+    pub class: CString,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// last password change
+    pub change: libc::time_t,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// expiration time of account
+    pub expire: libc::time_t
+}
+
+impl From<libc::passwd> for Passwd {
+    fn from(pw: libc::passwd) -> Passwd {
+        unsafe {
+            Passwd {
+                name: CString::new(CStr::from_ptr(pw.pw_name).to_bytes()).unwrap(),
```

Comment:
I wonder if it is smart to use `CStrings` here (as opposed to Strings).
Does the library use `CStrings` in any other places?

Reply:
I took a quick search of the code, and it does seem that we are using `String` and not `CString` as an interface. Please switch to `String`.

Ref:

* https://github.com/nix-rust/nix/blob/a00bd10dc47856418b5ebad62a16035577ebfe65/src/ifaddrs.rs#L21
* https://github.com/nix-rust/nix/blob/ebf75050f2f2726808308f76253f27c97aa6db15/src/pty.rs#L146
* https://github.com/nix-rust/nix/blob/ebf75050f2f2726808308f76253f27c97aa6db15/src/pty.rs#L167
* et cetera

Thanks!

Reply:
Thank you very much for your review. I did have what I believed to be a good reason to use CStrings. If this function is ran on a non UTF8 system, the fields, especially `gecos`, can be in a non UTF8 encoding such as SHIFTIJIS. `from_utf8_lossy` will cause data loss. Should String still be used in light of this?

Reply:
Does Nix support any operating systems that run in non-UTF8 mode?

Reply:
Yes, many. FreeBSD, Linux etc can all use `/etc/passwd` files that are in the user's non-UTF8 locale. I hate to keep mentioning SHIFT-JIS, but because in my day job I'm dealing with a lot of SJIS data and SJIS systems, I am accutely aware of this problem. Japanese sysadmins, even in 2018, in many cases still prefer SJIS because it is byte for byte around 2/3 to 1/2 the size of UTF-8. So that means an uncompressed database of Japanese text data is double the size if it's UTF-8. Therefore, many many Unix systems in Japan use SJIS everywhere, including the passwd and group files.